docs/dev/relay_add_pass.rst

```diff
@@ -0,0 +1,318 @@
+``CallNode``. In the following paragraphs, we explain the roles of each in
+the pass.
+
+.. code:: c
+
+    Expr VisitExpr_(const LetNode* op) final {
+      Expr value = this->Mutate(op->value);
+      if (value.as<ConstantNode>()) {
+        memo_[op->var] = value;
+        return this->Mutate(op->body);
+      } else {
+        Var var = Downcast<Var>(this->Mutate(op->var));
+        Expr body = this->Mutate(op->body);
+        if (var.same_as(op->var) &&
+            value.same_as(op->value) &&
+            body.same_as(op->body)) {
+          return GetRef<Expr>(op);
+        } else {
+          return LetNode::make(var, value, body);
+        }
+      }
+    }
+
+In the ``LetNode`` case, we first attempt to const-fold the value being bound
+in the expression. If we can, then we populate ``memo_`` and return the
+result of visiting the body---essentially, propagating the bound value to its
+use sites in the body. If we can't const-fold the bound value, we mimic the
+default implementation.
+
+.. code:: c
+
+    Expr VisitExpr_(const TupleGetItemNode* op) final {
+      Expr res = ExprMutator::VisitExpr_(op);
+      op = res.as<TupleGetItemNode>();
+      if (const auto* tuple = op->tuple.as<TupleNode>()) {
+        return tuple->fields[op->index];
+      } else {
+        return res;
+      }
+    }
+
+In the ``TupleItemGetNode`` case, we check if ``op->tuple`` field is a
+``TupleNode``. If so, we replace the tuple get with the field of the tuple
+pointed to by ``op->index``. The reason we need to check is because
+``op->tuple`` might evaluate to a tuple, without itself being a tuple.
+
+.. code:: c
+
+    Expr VisitExpr_(const CallNode* call) final {
+      static auto op_stateful = Op::GetAttr<TOpIsStateful>("TOpIsStateful");
+      Expr res = ExprMutator::VisitExpr_(call);
+      call = res.as<CallNode>();
+      // We don't constant fold function with zero arguments.
+      // This is a heuristic that is useful.
+      // For example it is harmful to fold ones(shape=(4, 5)).
+      if (call->args.size() == 0) return res;
+      const OpNode* op = call->op.as<OpNode>();
+      if (op == nullptr) return res;
+      // skip stateful ops.
+      if (op_stateful.get(GetRef<Op>(op), false)) return res;
+      bool all_const_args = true;
+      for (Expr arg : call->args) {
+        if (!checker_.Check(arg)) {
+          all_const_args = false;
+        }
+      }
+      if (all_const_args) {
+        return ConstEvaluate(res);
+      } else {
+        return res;
+      }
+    }
+
+In the ``CallNode`` case, we first use the ``VisitExpr_`` of ``ExprMutator``
+to visit the call, which const-folds all of the call parameters. Then we
+evaluate the call only if all of the arguments are constant (using
+``ConstantChecker``). Evaluating the call produces a value, so we use a
+helper method ``ValueToExpr`` to allow us to place the evaluated expression
+back into the AST.
+
+TODO: Why does the ``CallNode`` case call ``VisitExpr_`` and not
```

Comment:
ExprMutator::VisitExpr_(call) use the default implementation for call, which is just recurse
ExprMutator::VisitExpr(call) do dispatch and reach here again, which is infinite loop